### PR TITLE
Composite Networks

### PR DESCRIPTION
**Pull request by jrzkaminski**
Added:

1. Composite Network Builder (duplicated fragments from evo will be eliminated in the future updates)
2. Composite Discrete and Continuous nodes, they basically only carry information about models hooked on them
3. Composite Network, it is fully capable of learning sampling and predicting with common network interface
4. Documentation on everything, some minor fixes too
5. Tests covering composite networks
6. Updated documentation for evolutionary method of learning classic BNs

### CODE REVIEW

File: bamt/builders/composite_builder.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        self,
+        descriptor: Dict[str, Dict[str, str]],
+        regressor: Optional[object] = None,
+    ):
+        super().__init__(descriptor, regressor)
+
+        # Notice that vertices are used only by Builders
+        self.vertices = []
+
+        # LEVEL 1: Define a general type of node: Discrete or Gaussian
```
Comment (Roman223) [inline bamt/builders/composite_builder.py:47]: LEVEL 1: Define a general type of node: discrete or continuous

File: bamt/builders/evo_builder.py [hunk 2]
```diff
@@ -76,6 +77,10 @@ def __init__(
         self.default_max_arity = 100
         self.default_max_depth = 100
         self.default_timeout = 180
+        self.default_num_of_generations = 50
+        self.default_early_stopping_iterations = 50
+        self.verbose = (True,)
```
Comment (Roman223) [inline bamt/builders/evo_builder.py:82]: What is tuple here for?

File: bamt/networks/base.py [hunk 3]
```diff
@@ -91,7 +93,7 @@ def add_nodes(self, descriptor: Dict[str, Dict[str, str]]):
         descriptor: dict with types and signs of nodes
         """
         if not self.validate(descriptor=descriptor):
-            if not self.type == "Hybrid":
+            if not self.type == "Hybrid" or not self.type == "Composite":
```
Comment (Roman223) [inline bamt/networks/base.py:96]: Has CompositeBN got `self.validate` method? Can't find it.. It seem, all method of learning were overwritten

File: bamt/utils/composite_utils/MLUtils.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            "lgbmreg": "LGBMRegressor",
+            "catboostreg": "CatBoostRegressor",
+            "xgboost": "XGBClassifier",
+            "logit": "LogisticRegression",
+            "bernb": "BernoulliNB",
+            "multinb": "MultinomialNB",
+            "dt": "DecisionTreeClassifier",
+            "rf": "RandomForestClassifier",
+            "mlp": "MLPClassifier",
+            "lgbm": "LGBMClassifier",
```
Comment (Roman223) [inline bamt/utils/composite_utils/MLUtils.py:50]: Will we leave `lgbm`?

File: bamt/builders/evo_builder.py (hunk 2 above)
Comment (jrzkaminski) [reply bamt/builders/evo_builder.py:82]: That's what black did. It's going to make it tuple in any case

File: bamt/utils/composite_utils/MLUtils.py (hunk 4 above)
Comment (jrzkaminski) [reply bamt/utils/composite_utils/MLUtils.py:50]: Yep, but it's debateable. I'll comment it out for now

File: bamt/networks/base.py (hunk 3 above)
Comment (jrzkaminski) [reply bamt/networks/base.py:96]: I've added it just in case, for the future

File: bamt/builders/composite_builder.py (hunk 1 above)
Comment (jrzkaminski) [reply bamt/builders/composite_builder.py:47]: fixed

**Review by Roman223 — CHANGES_REQUESTED** [review]

File: bamt/networks/composite_bn.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                node.type = re.sub(
+                    r"\([\s\S]*\)", f"({type(node.classifier).__name__})", node.type
+                )
+            else:
+                continue
+
+    def set_regressor(self, regressors: Dict[str, object]):
+        """
+        Set regressor for gaussian nodes.
+        classifiers: dict with node_name and Classifier
```
Comment (Roman223) [inline bamt/networks/composite_bn.py:103]: typo in docstring: classifiers -> regressors

**Review by Roman223 — APPROVED** [review]

**Review by Anaxagor — CHANGES_REQUESTED** [review]
А добавь ещё тест, где ты для обучения композитной модели передаёшь только ген. операторы, которые не меняют структуру, а только меняют parent_model, и задай стартовую структуру для обучения, а на выходе проверь, что после обучения структура не поменялась.

**Review by Roman223 — APPROVED** [review]

**Pull request merged**